src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,134 @@
+using namespace chip::Crypto;
+using namespace chip::SetupPayload;
+using namespace chip::Encoding::LittleEndian;
+
+AdditionalDataPayloadGenerator::AdditionalDataPayloadGenerator(uint16_t rotationCounter, char * serialNumberBuffer,
+                                                               size_t serialNumberBufferSize) :
+    mRotationCounter(rotationCounter),
+    mSerialNumberBuffer(serialNumberBuffer), mSerialNumberBufferSize(serialNumberBufferSize)
+{
+    mHashInput =
+        reinterpret_cast<char *>(chip::Platform::MemoryAlloc(ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + serialNumberBufferSize));
+}
+
+AdditionalDataPayloadGenerator::~AdditionalDataPayloadGenerator()
+{
+    if (mHashInput != nullptr)
+    {
+        chip::Platform::MemoryFree(mHashInput);
+        mHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle,
+                                                                         bool enableRotatingDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(rotatingDeviceIdBuffer, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char * rotatingDeviceIdBuffer,
+                                                                    size_t & rotatingDeviceIdBufferSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char rotationCounterBuf[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    char outputBuffer[ROTATING_DEVICE_ID_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter hashInputBufferWriter(reinterpret_cast<unsigned char *>(mHashInput),
+                                       ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize);
+    BufferWriter outputBuferWriter(reinterpret_cast<unsigned char *>(&outputBuffer[0]),
+                                   sizeof(outputBuffer) / sizeof(outputBuffer[0]));
+
+    snprintf(rotationCounterBuf, sizeof(rotationCounterBuf), "%09u", mRotationCounter);
+    hashInputBufferWriter.Put(rotationCounterBuf, ROTATING_DEVICE_ID_COUNTER_STR_LENGTH);
```

Comment:
There's some assumption here about how the snprintf is touching all the bytes up to ROTATING_DEVICE_ID_COUNTER_STR_LENGTH, no?  What enforces that?

Or put another way, maybe the length should be the actual length (that is, 9), the buffer size should use that length + 1, and the snprintf format string should include `STR(ROTATING_DEVICE_ID_COUNTER_STR_LENGTH)` to make sure that all these things obviously match up?

Here `STR` would be defined as:
```
#define STR_IMPL_(x) #x
#define STR(x) STR_IMPL_(x)  // expands 'x' if it's a macro.
```

Reply:
rotationCounterBuf has the rotation counter and the rest of characters are filled by zeros, that's what the %09u is for.